### PR TITLE
HB-05

### PR DESCRIPTION
**Pull request by Max-Pochepets**
Task completed:

- Model, Service and Dao layers for User implemented
- Authentication service implemented
- Password hashing implemented

### CODE REVIEW

File: src/main/java/cinema/shop/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        User user = null;
```
Comment (Mandybur800) [inline src/main/java/cinema/shop/dao/impl/UserDaoImpl.java]: redundant line

File: src/main/java/cinema/shop/security/AuthenticationService.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package cinema.shop.security;
+
+import cinema.shop.lib.exception.AuthenticationException;
+import cinema.shop.model.User;
+
+public interface AuthenticationService {
+    User login(String email, String password) throws AuthenticationException;
+
+    User register(String email, String password) throws AuthenticationException;
```
Comment (Mandybur800) [inline src/main/java/cinema/shop/security/AuthenticationService.java]: why are you throw `AuthenticationException` in `registration()`?

File: src/main/java/cinema/shop/security/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userByEmail = userService.findByEmail(email);
+        if (userByEmail.isEmpty()) {
+            throw new AuthenticationException("Wrong credentials. Please try again.");
+        }
```
Comment (Mandybur800) [inline src/main/java/cinema/shop/security/AuthenticationServiceImpl.java]: you can add constrain `nullable = false` in User model

**Review by Mandybur800 — CHANGES_REQUESTED** [review]

**Review by Mandybur800 — APPROVED** [review]

File: src/main/java/cinema/shop/security/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        User user = userByEmail.get();
+        byte[] salt = user.getSalt();
+        if (!HashUtil.hashPassword(password, salt).equals(user.getPassword())) {
+            throw new AuthenticationException("Wrong credentials. Please try again.");
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/shop/security/AuthenticationServiceImpl.java]: let's combine these two checks so we are throwing exception only in one place

File: src/main/java/cinema/shop/security/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        return user;
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RegistrationException("Email " + email
+                    + " is already being used. Please choose another.");
+        }
```
Comment (dmytro-hryhoruk) [inline src/main/java/cinema/shop/security/AuthenticationServiceImpl.java]: email field has a unique constraint.  So exception will be thrown anyway. We can remove this check

**Review by dmytro-hryhoruk — CHANGES_REQUESTED** [review]

**Review by dmytro-hryhoruk — APPROVED** [review]

**Pull request merged**